lib/aliquot/payment.rb

```diff
@@ -0,0 +1,134 @@
+    # merchant_id::   Google Pay merchant ID ("merchant:<SOMETHING>")
+    # logger::        The logger to use. Default: Logger.new($stdout)
+    # signing_keys::  Formatted list of signing keys used to sign token contents.
+    #                 Otherwise a thread continuously updating google signing
+    #                 keys will be started.
+    def initialize(token_string, shared_secret, merchant_id, logger: Logger.new($stdout), signing_keys: nil)
+      Aliquot.start_key_updater(logger) if $key_updater_thread.nil? && signing_keys.nil?
+
+      validation = Aliquot::Validator::Token.new(JSON.parse(token_string))
+      validation.validate
+
+      @token = validation.output
+
+      @shared_secret = shared_secret
+      @merchant_id   = merchant_id
+      @signing_keys  = signing_keys
+    end
+
+    ##
+    # Validate and decrypt the token.
+    def process
+      unless valid_protocol_version?
+        raise Error, 'only ECv1 protocolVersion is supported'
+      end
+
+      unless valid_signature?
+        raise InvalidSignatureError
+      end
+
+      validator = Aliquot::Validator::SignedMessage.new(JSON.parse(@token['signedMessage']))
+      validator.validate
+      signed_message = validator.output
+
+      aes_key, mac_key = derive_keys(signed_message['ephemeralPublicKey'], @shared_secret, 'Google')
+
+      unless self.class.valid_mac?(mac_key, signed_message['encryptedMessage'], signed_message['tag'])
+        raise InvalidMacError
+      end
+
+      @message = JSON.parse(decrypt(aes_key, signed_message['encryptedMessage']))
+
+      Aliquot::Validator::EncryptedMessageValidator.new(@message).validate
+
+      raise ExpiredException if expired?
+
+      @message
+    end
+
+    def protocol_version
+      @token['protocolVersion']
+    end
+
+    def valid_protocol_version?
+      protocol_version == 'ECv1'
+    end
+
+    ##
+    # Check if the token is expired, according to the messageExpiration included
+    # in the token.
+    def expired?
+      @message['messageExpiration'].to_f / 1000.0 <= Time.now.to_f
+    end
+
+    def valid_signature?
+      signed_string = ['Google', @merchant_id, protocol_version, @token['signedMessage']].map do |str|
+        [str.length].pack('V') + str
+      end.join
+
+      keys = JSON.parse(signing_keys)['keys']
+      # Check if signature was performed with any possible key.
+      keys.map do |e|
+        next if e['protocolVersion'] != protocol_version
+
+        ec = OpenSSL::PKey::EC.new(Base64.strict_decode64(e['keyValue']))
+        d  = OpenSSL::Digest::SHA256.new
+        ec.verify(d, Base64.strict_decode64(@token['signature']), signed_string)
+      end.any?
+    end
+
+    private
+
+    def derive_keys(ephemeral_public_key, shared_secret, info)
+      ikm = Base64.strict_decode64(ephemeral_public_key) +
+            Base64.strict_decode64(shared_secret)
+      hbytes = HKDF.new(ikm, algorithm: 'SHA256', info: info).next_bytes(32)
```

Comment:
What does the `h` in `hbytes` stand for?

Reply:
`hash`, but `key` is obviously more correct.